### PR TITLE
SNOW-733213: fixing the `auto_create_table` behavior with `overwrite` option

### LINKED ISSUE #1413: SNOW-733213: Add ability for write_pandas to truncate table instead of drop table
Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using?

   3.9.0

2. What operating system and processor architecture are you using?

macOS-13.1-x86_64-i386-64bit

3. What are the component versions in the environment (`pip freeze`)?

snowflake-connector-python==2.9.0

4. What did you do?

Using write_pandas with overwrite=true, the table on Snowflake side is dropped and then recreated.  This was decided in this git issue:

https://github.com/snowflakedb/snowflake-connector-python/issues/1175

The decision was made in order to maintain the dataframe schema.  However, drop tables loses grants on the table.

5. What did you expect to see?

We would like to have an option that truncates the table instead of dropping the table in order to maintain grants.

### PR DESCRIPTION
**Pull request by andreqaugusto**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #1413 

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [x] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

  We are modifying the code in some places to `TRUNCATE` the table as the docstring says it should be.

### CODE REVIEW

File: src/snowflake/connector/pandas_tools.py [hunk 1]
```diff
@@ -389,7 +394,7 @@ def drop_object(name: str, object_type: str) -> None:
         target_table_location = build_location_helper(
             database,
             schema,
-            random_string() if overwrite else table_name,
+            random_string() if (overwrite and auto_create_table) else table_name,
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/pandas_tools.py:397]: This seems like an unrelated bug fix to what you describe in your PR description. Could you write a test for this?

File: src/snowflake/connector/pandas_tools.py [hunk 2]
```diff
@@ -362,6 +362,11 @@ def drop_object(name: str, object_type: str) -> None:
         logger.debug(f"dropping {object_type} with '{drop_sql}'")
         cursor.execute(drop_sql, _is_internal=True)
 
+    def truncate_table(name: str) -> None:
+        truncate_sql = f"TRUNCATE TABLE IF EXISTS {name} /* Python:snowflake.connector.pandas_tools.write_pandas() */"
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/pandas_tools.py:366]: I believe that it'd be quite an unexpected situation if it didn't exist. 
```suggestion
        truncate_sql = f"TRUNCATE TABLE {name} /* Python:snowflake.connector.pandas_tools.write_pandas() */"
```

File: src/snowflake/connector/pandas_tools.py [hunk 3]
```diff
@@ -362,6 +362,11 @@ def drop_object(name: str, object_type: str) -> None:
         logger.debug(f"dropping {object_type} with '{drop_sql}'")
         cursor.execute(drop_sql, _is_internal=True)
 
+    def truncate_table(name: str) -> None:
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/pandas_tools.py:365]: Since this function is only called from 1 spot then let's not declare it as a helper method.

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: src/snowflake/connector/pandas_tools.py (hunk 1 above)
Comment (andreqaugusto) [reply src/snowflake/connector/pandas_tools.py:392]: it is related to the bug fix (and the bug itself), but nonetheless I wrote a new test for it (since I hadn't written earlier)

**Comment by andreqaugusto** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by andreqaugusto** [comment]
recheck

**Comment by andreqaugusto** [comment]
hey! just bumping to see if we can get this through (it is a feature that I want to use at my work)

**Review by sfc-gh-mkeller — APPROVED** [review]
:ship: :ship:

**Comment by sfc-gh-mkeller** [comment]
It looks like in the test that you added [this](https://github.com/andreqaugusto/snowflake-connector-python/blob/b2927c8d81702f4d732c43d886c13304db0f1f94/src/snowflake/connector/pandas_tools.py#L453) gets called. So that the table will not exist when you will want to select from it.
However; I think even if the table wasn't dropped the truncate will empty the table, so your assertion will fail anyways.

**Comment by sfc-gh-aalam** [comment]
@andreqaugusto I merged your changes with a minor test update in this PR: https://github.com/snowflakedb/snowflake-connector-python/pull/1833
